### PR TITLE
add storage layer interfaces

### PR DESCRIPTION
**Pull request by whywhy36**
Only for review, @easeway please help review the interfaces.

### CODE REVIEW

**Comment by whywhy36** [comment]
check github bot notification

**Comment by easeway** [comment]
I see in your model, there's a concept of Queue. While in decentralized model, there's no need of a Queue. What's need from the store is basically a Query and an Update. The workers picks up tasks by querying tasks by state, and then compete to get a task. Once owned the task, the worker is able to update the state of the task.

For example, the following logic in a worker:

```
    tasks := store.QueryTasks("state", TaskPending)
    for _, task := range tasks {
        err := store.ClaimTask(task.ID)
        if err == nil { // successfully claimed the ownership
            RunTask(task)
            break
        }
    }
```

And the design simplifies the submission of a job/task. It's as simple as inserting a new object

```
    store.AddEntity(job)
    store.AddEntity(task)
    ... etc
```

Similarly, some house keeping routines can easily recover/cleanup tasks

```
    tasks := store.QueryTimeoutTasks()
    for _, task := range tasks {
        if store.ClaimTask(task) == nil {
            task.Errors = append(task.Errors, TaskTimeoutError("timeout"))
            task.Retries ++
            task.State = TaskPending
            store.UpdateTask(task)
        }
    }
```

The methods `Query`,  `QueryTimeoutTasks` etc just come out from my mind. They need better modeling and re-design. But the general idea behind is a eventually consistent model that workers run in a competing way. The query part is most important in the store to support different scenarios: run tasks, clean up, retry, rollback, prioritize (future functionality) etc.

**Comment by whywhy36** [comment]
@easeway  yes, the `queue` concept could be hidden from the direct interface. But we may need to let tasks processed in order otherwise some starvation may happen. I think `queue` concept could be useful in this case. `queue` should be the native adapter interface for all persisters.

For the two target back-end persisters `etcd` and `consul`, i knew that `etcd` supports ordering dictionary but still not quite sure whether `consul` supports it or not. The ordering thing could be approached by putting timestamp or some ordering Ids to the item's value if there's no native support for ordering.

**Comment by easeway** [comment]
ordering is not a requirement, as all tasks are submitted dynamically. the idea is once a task is submitted, it's ready to run. there's no order. but we may consider the fairness. as older tasks should take higher priority, relatively, but not strictly. I'm not sure if timestamps (like created_at, updated_at) can be utilized to sort the keys in etcd/consul.

I believe without building a queue simplifies the design and implementation.

**Comment by whywhy36** [comment]
@easeway @chempher Please help review the updated interfaces.

I split the one store interface to 3 separate ones. Each for one type of data structure (key-value, queue style dictionary and CAS one).

The basic usage should like 

```
var kvStore := KVStoreImpl.new("tasks")
var queueStore := QueueStoreImpl.new("active")
...
kvStore.Put("task_id", ...)
queueStore.Enqueue("task_id", ...)
...
```

The first level key name would be passed in as the parameter of store constructor. This will ease the further usage and simplify the method signature.

As this interface is the contract between Strategy and Store, we should agree on it first. I am working on the native adapter interfaces. Would send the pull request together with the confirmed contract store interfaces.

File: storage/store.go [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    // update a object
+    Put(key string, value string) error
+    Get(key string) (string, error)
+    Del(key string) error
+
+    Contains(key string) bool
+}
+
+type CASStore interface {
+    Claim(key string) (string, error)
```
Comment (easeway) [inline storage/store.go]: where is the owner id? I thought it should be `Claim(key, ownerID string)`, and what's the returned `string` for? And after updates complete, how can the owner unclaim the ownership? What in my mind is like:

```
type Claim interface {
    Object() string        // returns owned object id
    Owner() string        // returns owner id
    Dismiss()                // the claim finished, and the object can be returned
}

type Store interface {
    ...
    Claim(id, ownerID string) (Claim, error) // if returned claim is nil, claim failed
}
```

File: storage/store.go [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    Get(key string) (string, error)
+    Del(key string) error
+    Contains(key string) bool
+}
+
+type QueueStore interface {
+    // append new object
+    Enqueue(key string, value string) error
+    Peek() (string, error)
+    Dequeue() (string, error)
```
Comment (easeway) [inline storage/store.go]: Enqueue/Dequeue indicates strict ordering. Can we use other words to eliminate the confusing?

File: storage/store.go [hunk 3]
```diff
@@ -0,0 +1,28 @@
+package storage
+
+type KVStore interface {
+    // key value
+    Put(key string, value string) error
+    Get(key string) (string, error)
+    Del(key string) error
+    Contains(key string) bool
```
Comment (easeway) [inline storage/store.go]: What we need is hierarchical. Starting from top-level `tasks` and then a directory/container `task-id`, and then a map like: `content` = task detailed information, `stats` = task runtime information, `cancellation` = cancellation request etc. How does this hierarchy reflected in the interface?

File: storage/store.go [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+}
+
+type QueueStore interface {
+    // append new object
+    Enqueue(key string, value string) error
+    Peek() (string, error)
```
Comment (easeway) [inline storage/store.go]: Not sure if we need peek.

File: storage/store.go [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+    Dequeue() (string, error)
+
```
Comment (easeway) [inline storage/store.go]: Regarding the queue store, the actual usage pattern is not like using a queue. What we need are only:
- put an id in
- get a list of limited number of ids which is relatively older
- delete an id

So, using queue pattern looks not very matching.

**Comment by easeway** [comment]
Overall, two more comments:
1. please move these into `jobs` package, we are not ready for creating a separate store package which is standalone.
2. please update `jobs.Store` interface to contain high-level methods. low-level adapter interface can come with a different name (or even internal name). The high-level methods should cover:
   - hierarchical K/V store, like a file system
   - some sort of queuing service (see my comments above)
   - owership claim

**Pull request closed without merging**